File: packages/ogdf.js/src/layouts/energybased/gem.js
import { PARAMETER_TYPE } from '../../utils/parameters'
import createLayout from '../creater'

const NAME = 'GEM'
const OUR_PARAMETER_DEFINITIONS = {
    useWorker: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: false
    }
}

const ORIGIN_PARAMETER_DEFINITIONS = {
    attractionFormula: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 1
    },
    desiredLength: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 5
    },
    gravitationalConstant: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 1 / 16
    },
    initialTemperature: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 10
    },
    maximalDisturbance: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 0
    },
    minDistCC: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 20
    },
    minimalTemperature: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 0.005
    },
    numberOfRounds: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 20000
    },
    oscillationAngle: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: Math.PI / 2
    },
    oscillationSensitivity: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 0.3
    },
    pageRatio: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 1
    },
    rotationAngle: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: Math.PI / 3
    },
    rotationSensitivity: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 0.001
    }
}

// double* nodesX, double* nodesY
const ATTRIBUTES_DEFINITIONS = {
    node: [
        {
            name: 'nodesX',
            mapper: (node) => ('x' in node ? node.x : 0),
            type: PARAMETER_TYPE.DOUBLE
        },
        {
            name: 'nodesY',
            mapper: (node) => ('y' in node ? node.y : 0),
            type: PARAMETER_TYPE.DOUBLE
        }
    ],
    link: [],
    sequence: ['nodesX', 'nodesY']
}

const gem = createLayout(
    NAME,
    OUR_PARAMETER_DEFINITIONS,
    ORIGIN_PARAMETER_DEFINITIONS,
    ATTRIBUTES_DEFINITIONS
)
export default gem

File: packages/ogdf.js/src/entry/energybased/gem.cpp
#include <ogdf/energybased/GEMLayout.h>
#include "../main.h"

EM_PORT_API(float *) GEM(int node_num, int link_num, int* source, int* target, double* nodesX, double* nodesY, int attractionFormula, double desiredLength, double gravitationalConstant, double initialTemperature, double maximalDisturbance, double minDistCC, double minimalTemperature, int numberOfRounds, double oscillationAngle, double oscillationSensitivity, double pageRatio, double rotationAngle, double rotationSensitivity){
	node* nodes;
	Graph G;
	GraphAttributes GA(G, GraphAttributes::nodeGraphics | GraphAttributes::edgeGraphics | GraphAttributes::edgeDoubleWeight);

	nodes = new node[node_num];
	for (int i = 0; i < node_num; i++){
		nodes[i] = G.newNode();
		GA.x(nodes[i]) = nodesX[i];
		GA.y(nodes[i]) = nodesY[i];
	}

	edge e;

	for (int i = 0; i < link_num; i++) {
		e = G.newEdge(nodes[source[i]], nodes[target[i]]);
		GA.bends(e);
	}
	//LayoutModule
	GEMLayout *model = new GEMLayout();

	model->attractionFormula(attractionFormula);
	model->desiredLength(desiredLength);
	model->gravitationalConstant(gravitationalConstant);
	model->initialTemperature(initialTemperature);
	model->maximalDisturbance(maximalDisturbance);
	model->minDistCC(minDistCC);
	model->minimalTemperature(minimalTemperature);
	model->numberOfRounds(numberOfRounds);
	model->oscillationAngle(oscillationAngle);
	model->oscillationSensitivity(oscillationSensitivity);
	model->pageRatio(pageRatio);
	model->rotationAngle(rotationAngle);
	model->rotationSensitivity(rotationSensitivity);

    model->call(GA);

	float* re = (float*)malloc(node_num * 2 * 4);
    for(int i = 0; i < node_num; ++i) {
        re[i * 2] = GA.x(nodes[i]);
        re[i * 2 + 1] = GA.y(nodes[i]);
    }

    return re;
}
